plugins/elanfp/fu-elanfp-usb-device.c

```diff
@@ -0,0 +1,564 @@
+#define FIRMWARE_UPDATE_ERROR_INVALID	   0x0B
+
+#define REPORT_ID_FW_VERSION_FEATURE 0x20
+#define REPORT_ID_OFFER_COMMAND	     0x25
+#define REPORT_ID_OFFER_RESPONSE     0x25
+#define REPORT_ID_PAYLOAD_COMMAND    0x20
+#define REPORT_ID_PAYLOAD_RESPONSE   0x22
+
+#define REQTYPE_GET_VERSION 0xC1
+#define REQTYPE_COMMAND	    0x41
+
+struct _FuElanfpUsbDevice {
+	FuUsbDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuElanfpUsbDevice, fu_elanfp_usb_device, FU_TYPE_USB_DEVICE)
+
+// Communication
+gboolean
+iapSendCommand(GUsbDevice *usb_device,
+	       guint8 reqType,
+	       guint8 request,
+	       guint8 *pbuff,
+	       gsize len,
+	       GError **error)
+{
+	gsize actual = 0;
+
+	if (pbuff == NULL || len == 0)
+		return FALSE;
+
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   request, // request
+					   0x00,    // value
+					   0x00,    // index
+					   pbuff,
+					   len,
+					   &actual,
+					   CTRL_SEND_TIMEOUT_MS,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "failed to send command: ");
+		return FALSE;
+	}
+
+	if (actual != len) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "send length (%d) is not match with the request (%d)",
+			    (gint32)actual,
+			    (gint32)len);
+
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+iapRecvStatus(GUsbDevice *usb_device, guint8 *pbuff, gsize len, GError **error)
+{
+	gsize actual = 0;
+
+	if (pbuff == NULL || len == 0)
+		return FALSE;
+
+	if (!g_usb_device_bulk_transfer(usb_device,
+					ELAN_EP_CMD_IN,
+					pbuff,
+					len,
+					&actual,
+					BULK_RECV_TIMEOUT_MS,
+					NULL,
+					error)) {
+		g_prefix_error(error, "failed to received status: ");
+		return FALSE;
+	}
+
+	if (actual != len) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "received length (%d) is not match with the request (%d)",
+			    (gint32)actual,
+			    (gint32)len);
+
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+gboolean
+runIapProcess(FuElanfpUsbDevice *self, GBytes *fw, GError **error)
+{
+	const guint8 *pbinary = NULL;
+	guint8 databuf[61] = {0};
+	guint8 recvbuf[17] = {0};
+	S2FFILE s2ffile;
+	PAYLOAD_HEADER *ppayload_header = NULL;
+	guint32 payloadheader_length = 5;
+	guint32 pkg_index = 0;
+	guint32 payload_offset = 0;
+	gsize binary_size;
+	gboolean result = TRUE;
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+
+	if (fw == NULL)
+		return FALSE;
+
+	pbinary = g_bytes_get_data(fw, &binary_size);
+
+	if (pbinary == NULL) {
+		// g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "device write fw - binary
+		// buffer is null !!\n");
+		return FALSE;
+	}
+
+	if (!binaryVerify(pbinary, binary_size, &s2ffile, error)) {
+		// g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE, "device write fw -
+		// binary verify fail !!\n");
```

Comment:
g_prefix_error

Reply:
Fixed in commit 5af9c23594ac261d85fbf53fca02de08a5ef0150.